### PR TITLE
Task/11371 - populate attribute

### PR DESCRIPTION
**Pull request by kenvac**
Add support to populate field based on other field value

### CODE REVIEW

File: src/core/containers.py [hunk 1]
```diff
@@ -34,3 +34,25 @@ def __init__(self, flow_name, task_name):
...
         return {"flow": self.flow, "task": self.task}
+
+
+class Populate(Container):
+    __slots__ = ["path", "validators", "value"]
+
+    def __init__(self, path="", validators="", value="", **kwargs):
+        super().__init__(**kwargs)
+        self.path = path
+        self.validators = validators
```
Comment (nocturnalastro) [inline src/core/containers.py:46]: Going to need a mechanism like the components to extract the validators and we will have to add to the input component
```Input.get_validators``` func to extract it.

File: src/core/containers.py [hunk 2]
```diff
@@ -34,3 +34,25 @@ def __init__(self, flow_name, task_name):
...
+        self.value = value
+
+    def as_dict(self):
+        retval = {}
+        if self.path:
+            retval.update(path=self.path)
+            if self.value:
+                raise Exception("'value' and 'path' attribute cannot be used together")
```
Comment (nocturnalastro) [inline src/core/containers.py]: Make an InvalidAguments exception for this so we have a specific type also do the check at __init__ so we can raise it earilier.

File: src/core/containers.py [hunk 3]
```diff
@@ -34,3 +34,25 @@ def __init__(self, flow_name, task_name):
...
+        if self.value:
+            retval.update(value=self.value)
+        if self.validators:
```
Comment (nocturnalastro) [inline src/core/containers.py]: validators should be a required field as it how we decide when to populate the field

File: src/core/containers.py [hunk 4]
```diff
@@ -34,3 +34,25 @@ def __init__(self, flow_name, task_name):
 
     def as_dict(self):
         return {"flow": self.flow, "task": self.task}
+
+
+class Populate(Container):
+    __slots__ = ["path", "validators", "value"]
+
+    def __init__(self, path="", validators="", value="", **kwargs):
```
Comment (nocturnalastro) [inline src/core/containers.py]: Use `None` instead of empty strings for optional arguments.

**Review by nocturnalastro — CHANGES_REQUESTED** [review]

File: src/core/containers.py (hunk 4 above)
Comment (kenvac) [reply src/core/containers.py]: Done

File: src/core/containers.py (hunk 3 above)
Comment (kenvac) [reply src/core/containers.py]: Done

File: src/core/containers.py (hunk 2 above)
Comment (kenvac) [reply src/core/containers.py]: Done

**Review by nocturnalastro — CHANGES_REQUESTED** [review]
Could you also update the jsonschema and write a quick test which validates against the schema,

File: src/core/components.py [hunk 5]
```diff
@@ -145,11 +148,15 @@ def get_base_component_dict(self):
...
             component["output"] = self.output
+        if self.populate:
+            component["populate"] = self.populate.as_dict()
         return component
 
     def get_validators(self):
         yield from super().get_validators()
         yield from self.validators
+        if self.populate:
+            yield from self.populate.validators
```
Comment (nocturnalastro) [inline src/core/components.py:159]: can you define a get_validators function on puplate and use that here that way if we need to change the internal structure of the populate container we can without having to change this :)

File: src/core/schema/components/input.json [hunk 6]
```diff
@@ -4,16 +4,85 @@
...
-    "type": { "type": "string", "enum": ["input"] },
-    "name": { "type": "string" },
-    "label": { "type": "string" },
-    "input_key": { "$ref": "../common.json#/definitions/jsonpath" },
-    "input_ref": { "$ref": "../common.json#/definitions/jsonpath" },
-    "output_ref": { "$ref": "../common.json#/definitions/jsonpath" },
-    "output": { "type": "object" },
-    "obscure": { "type": "boolean" },
-    "validators": { "type": "array", "items": [{ "type": "string" }] }
+    "type": {
```
Comment (nocturnalastro) [inline src/core/schema/components/input.json:44]: Can you put these back on to one-liners? It becomes much less readable like this.

File: src/core/schema/components/input.json [hunk 7]
```diff
@@ -4,16 +4,85 @@
   "title": "input",
   "description": "Workflows input component",
   "type": "object",
+  "definitions": {
+    "populate": {
```
Comment (nocturnalastro) [inline src/core/schema/components/input.json:8]: Can you move this into its own file that way we can test the populate object independatly of input

**Review by nocturnalastro — CHANGES_REQUESTED** [review]

**Comment by nocturnalastro** [comment]
Also I pushed a fix for the failing unit test please rebase :)

**Comment by nocturnalastro** [comment]
Ok so this was closed when I merged and deleted the target branch...seems like bad GitHub UX.

**Pull request closed without merging**